sycl/include/sycl/ext/oneapi/annotated_arg/annotated_arg.hpp

```diff
@@ -0,0 +1,286 @@
+  // annotated_arg object. The same property in `Props...` and `PropertyList2`
+  // must have the same property value.
+  template <typename T2, typename PropertyList2>
+  explicit annotated_arg(const annotated_arg<T2, PropertyList2> &other) noexcept
+      : obj(other.obj) {
+    static_assert(std::is_convertible<T2, T *>::value,
+                  "The underlying data type of the input annotated_arg is not "
+                  "compatible");
+
+    static_assert(
+        std::is_same<
+            property_list_t,
+            detail::merged_properties_t<property_list_t, PropertyList2>>::value,
+        "The constructed annotated_arg type must contain all the properties of "
+        "the input annotated_arg");
+  }
+
+  // Constructs an annotated_arg object from another annotated_arg object and a
+  // property list. The new property set is the union of property lists
+  // `PropertyListU` and `PropertyListV`. The same property in `PropertyListU`
+  // and `PropertyListV` must have the same property value.
+  template <typename T2, typename PropertyListU, typename PropertyListV>
+  explicit annotated_arg(const annotated_arg<T2, PropertyListU> &other,
+                         const PropertyListV &proplist) noexcept
+      : obj(other.obj) {
+    static_assert(std::is_convertible<T2, T *>::value,
+                  "The underlying data type of the input annotated_arg is not "
+                  "compatible");
+
+    static_assert(
+        std::is_same<property_list_t, detail::merged_properties_t<
+                                          PropertyListU, PropertyListV>>::value,
+        "The property list of constructed annotated_arg type must be the union "
+        "of the input property lists");
+  }
+
+  operator T *() noexcept { return obj; }
+  operator T *() const noexcept { return obj; }
+
+  T &operator[](std::ptrdiff_t idx) const noexcept { return obj[idx]; }
+
+  template <typename PropertyT> static constexpr bool has_property() {
+    return property_list_t::template has_property<PropertyT>();
+  }
+
+  template <typename PropertyT> static constexpr auto get_property() {
+    return property_list_t::template get_property<PropertyT>();
+  }
+};
+
+// Partial specialization for non-pointer type
+template <typename T, typename... Props>
+class __SYCL_SPECIAL_CLASS
+__SYCL_TYPE(annotated_arg) annotated_arg<T, detail::properties_t<Props...>> {
+  using property_list_t = detail::properties_t<Props...>;
+
+  template <typename T2, typename PropertyListT> friend class annotated_arg;
+
+  T obj;
+
+#ifdef __SYCL_DEVICE_ONLY__
+  void __init([[__sycl_detail__::add_ir_attributes_kernel_parameter(
+      detail::PropertyMetaInfo<Props>::name...,
+      detail::PropertyMetaInfo<Props>::value...)]] T _obj) {
+    obj = _obj;
+  }
+#endif
+
+public:
+  // T should be trivially copyable to be device-copyable
+  // static_assert(std::is_trivially_copyable<T>::value,
+  //               "Type T must be trivially copyable.");
```

Comment:
Can this be removed?

Reply:
This is actually a requirement in SYCL spec: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec::device.copyable

Reply:
By testing on cases where `T` is not device copyable (e.g. `T=std::pair<int,bool>`),  the compile runs through without capturing it.  So I think we should keep this check until the compiler can capture it

Reply:
Interesting. Could you please make it `is_device_copyable_v` instead then?